storage_v2/v1adapter/ptrace2model.go

```diff
@@ -0,0 +1,74 @@
+// Copyright (c) 2024 The Jaeger Authors.
+// SPDX-License-Identifier: Apache-2.0
+
+package v1adapter
+
+import (
+	"go.opentelemetry.io/collector/pdata/ptrace"
+
+	"github.com/jaegertracing/jaeger/model"
+	"github.com/jaegertracing/jaeger/pkg/iter"
+	"github.com/jaegertracing/jaeger/storage/spanstore"
+	otel2model "github.com/open-telemetry/opentelemetry-collector-contrib/pkg/translator/jaeger"
+)
+
+// PTracesSeq2ToModel consumes an iterator seqTrace. When necessary,
+// it groups spans from *consecutive* chunks of ptrace.Traces into a single model.Trace
+//
+// Returns nil, and spanstore.ErrTraceNotFound for empty iterators
+func PTracesSeq2ToModel(seqTrace iter.Seq2[[]ptrace.Traces, error]) ([]*model.Trace, error) {
+	var (
+		err         error
+		lastTraceID *model.TraceID
+		lastTrace   *model.Trace
+	)
+	jaegerTraces := []*model.Trace{}
+
+	seqTrace(func(otelTraces []ptrace.Traces, e error) bool {
+		if e != nil {
+			err = e
+			return false
+		}
+
+		for _, otelTrace := range otelTraces {
+			spans := modelSpansFromOtelTrace(otelTrace)
+			if len(spans) == 0 {
+				continue
+			}
+			currentTraceID := spans[0].TraceID
+			if lastTraceID != nil && *lastTraceID == currentTraceID {
+				lastTrace.Spans = append(lastTrace.Spans, spans...)
+			} else {
+				newTrace := &model.Trace{Spans: spans}
+				lastTraceID = &currentTraceID
+				lastTrace = newTrace
+				jaegerTraces = append(jaegerTraces, lastTrace)
+			}
+		}
+		return true
+	})
+
+	if err != nil {
+		return nil, err
+	}
+
+	if len(jaegerTraces) == 0 {
+		return nil, spanstore.ErrTraceNotFound
+	}
+	return jaegerTraces, nil
+}
+
+// modelSpansFromOtelTrace extracts spans from otel traces
+func modelSpansFromOtelTrace(otelTrace ptrace.Traces) []*model.Span {
+	spans := []*model.Span{}
+	batches := otel2model.ProtoFromTraces(otelTrace)
+	for _, batch := range batches {
+		for _, span := range batch.Spans {
+			span.Process = &model.Process{}
+			span.Process.ServiceName = batch.GetProcess().GetServiceName()
+			span.Process.Tags = batch.GetProcess().GetTags()
```

Comment:
what we usually do in this case is
```
if span.Process == nil {
  span.Process = batch.Process
}
```

Reply:
Alright...

I tried updating the span process everytime, regardless, but the comparator would pick up a difference I couldn't notice by looking at json data. But I'll try with this.

Reply:
reusing the Process from Batch avoids unnecessary memory allocations.

Reply:
Okay,

![Screenshot (83)](https://github.com/user-attachments/assets/82ecdedc-8351-4ecf-ae1c-c6854cfa6c83)
![Screenshot (84)](https://github.com/user-attachments/assets/669af168-7ea4-4f1e-b56b-a34d066ba2e9)

Reply:
I spent an hour debugging but couldn't find the root cause. I even wrote a simple test that does a roundtrip from model -> ptrace -> model where we can see that if `Process.Tags==[]` at the start it becomes `nil` after roundtrip. However, `pretty.Diff` did not complain about that, so I am not sure why it was complaining in the badger test (it complained exactly about that, that actual `Process.Tags == nil` while expected process had `[]`.